zaza/openstack/charm_tests/neutron/tests.py

```diff
@@ -29,6 +30,206 @@
+        logging.info(
+            'Setting verbose on neutron-api {}'.format(set_alternate))
+        self.restart_on_changed(
+            conf_file,
+            set_default,
+            set_alternate,
+            default_entry,
+            alternate_entry,
+            self.services,
+            pgrep_full=self.pgrep_full)
+
+    def test_910_pause_and_resume(self):
+        """Run pause and resume tests.
+
+        Pause service and check services are stopped then resume and check
+        they are started
+        """
+        with self.pause_resume(
+                self.services,
+                pgrep_full=self.pgrep_full):
+            logging.info("Testing pause resume")
+
+    def test_920_change_aa_profile(self):
+        """Test changing the Apparmor profile mode."""
+        services = ['neutron-openvswitch-agent',
+                    'neutron-dhcp-agent',
+                    'neutron-l3-agent',
+                    'neutron-metadata-agent',
+                    'neutron-metering-agent']
+
+        set_default = {'aa-profile-mode': 'disable'}
+        set_alternate = {'aa-profile-mode': 'complain'}
+
+        mtime = zaza.model.get_unit_time(
+            self.lead_unit,
+            model_name=self.model_name)
+        logging.debug('Remote unit timestamp {}'.format(mtime))
+
+        with self.config_change(set_default, set_alternate):
+            for unit in zaza.model.get_units('neutron-gateway',
+                                             model_name=self.model_name):
+                logging.info('Checking number of profiles in complain '
+                             'mode in {}'.format(unit.entity_id))
+                run = zaza.model.run_on_unit(
+                    unit.entity_id,
+                    'aa-status --complaining',
+                    model_name=self.model_name)
+                output = run['Stdout']
+                self.assertTrue(int(output) >= len(services))
+
+    @classmethod
+    def _get_services(cls):
+        """
+        Return the services expected in Neutron Gateway.
+
+        :returns: A list of services
+        :rtype: list[str]
+        """
+        services = ['neutron-dhcp-agent',
+                    'neutron-metadata-agent',
+                    'neutron-metering-agent',
+                    'neutron-openvswitch-agent']
+
+        trusty_icehouse = openstack_utils.get_os_release('trusty_icehouse')
+        xenial_newton = openstack_utils.get_os_release('xenial_newton')
+        bionic_train = openstack_utils.get_os_release('bionic_train')
+
+        if cls.current_os_release <= trusty_icehouse:
+            services.append('neutron-vpn-agent')
+        if cls.current_os_release < xenial_newton:
+            services.append('neutron-lbaas-agent')
+        if xenial_newton <= cls.current_os_release < bionic_train:
+            services.append('neutron-lbaasv2-agent')
+
+        return services
+
+
+class NeutronGatewaySecurityTest(test_utils.OpenStackBaseTest):
```

Comment:
Would it be possible to generalize this class with the existing SecurityTest class. Just some intelligence about the application name seems necessary.

Reply:
done - [commit](https://github.com/openstack-charmers/zaza-openstack-tests/pull/142/commits/79bb61243f25d3264ab81edfd9279036d930dd5d)

I considered doing this initially, but hesitated to touch another charm at the same time. Func-smoke is happy for both neutron-api and neutron-gateway with this commit (should essentially be a noop for neutron-api)